src/functional.rs

```diff
@@ -0,0 +1,98 @@
+//! Functional programming with [`Array`]s.
+// This is modeled after `generic-array::functional`
+// see: <https://docs.rs/generic-array/1.0.0/generic_array/functional/index.html>
+
+use crate::{Array, ArraySize};
+
+impl<T, U> Array<T, U>
+where
+    U: ArraySize,
+{
+    /// Maps an `Array` to another `Array`
+    ///
+    /// If the mapping function panics, any already initialized elements in the new array will
+    /// be dropped, AND any unused elements in the source array will also be dropped.
+    #[inline(always)]
+    pub fn map<O, F>(self, f: F) -> Array<O, U>
+    where
+        F: FnMut(T) -> O,
+    {
+        Mapper {
+            inner: self.into_iter(),
+            f,
+        }
+        .collect()
```

Comment:
```suggestion
        Self(self.0.map(f))
```

Reply:
that didn't work, I had to get an iterator instead.